File: backend/src/models/Pagamento.js
'use strict';

const { Model, DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  class Pagamento extends Model {
    static associate(models) {
      // Associação com Cliente
      Pagamento.belongsTo(models.Cliente, {
        foreignKey: 'cliente_id',
        as: 'cliente'
      });
      
      // Associação com User (quem criou o pagamento)
      Pagamento.belongsTo(models.User, {
        foreignKey: 'created_by',
        as: 'criador'
      });
    }
  }

  Pagamento.init({
    cliente_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: 'clientes',
        key: 'id'
      }
    },
    created_by: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id'
      }
    },
    mp_preference_id: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'ID da preferência no Mercado Pago'
    },
    mp_payment_id: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'ID do pagamento no Mercado Pago'
    },
    tipo: {
      type: DataTypes.ENUM('boleto', 'pix', 'cartao', 'universal'), // ✅ ADICIONADO 'universal'
      allowNull: false,
      defaultValue: 'universal' // ✅ MUDADO PARA UNIVERSAL
    },
    status: {
      type: DataTypes.ENUM('pendente', 'aprovado', 'rejeitado', 'cancelado', 'expirado', 'aguardando'), // ✅ ADICIONADO 'aguardando'
      allowNull: false,
      defaultValue: 'pendente'
    },
    titulo: {
      type: DataTypes.STRING,
      allowNull: false
    },
    descricao: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    valor: {
      type: DataTypes.STRING,
      allowNull: false,
      comment: 'Valor formatado como string (ex: 2.000,00)'
    },
    valor_numerico: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      comment: 'Valor numérico para cálculos'
    },
    // ✅ CAMPOS PARA PARCELAS
    parcelas: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 1,
      comment: 'Número de parcelas (1 = à vista)'
    },
    valor_parcela: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Valor de cada parcela formatado'
    },
    valor_parcela_numerico: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true,
      comment: 'Valor numérico de cada parcela'
    },
    // ✅ CAMPOS PARA CONTROLE DE ENVIOS
    whatsapp_enviado: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      comment: 'Se foi enviado para WhatsApp'
    },
    email_enviado: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      comment: 'Se foi enviado por email'
    },
    data_envio_whatsapp: {
      type: DataTypes.DATE,
      allowNull: true,
      comment: 'Data do envio do WhatsApp'
    },
    data_envio_email: {
      type: DataTypes.DATE,
      allowNull: true,
      comment: 'Data do envio do email'
    },
    data_vencimento: {
      type: DataTypes.DATE,
      allowNull: true
    },
    data_pagamento: {
      type: DataTypes.DATE,
      allowNull: true
    },
    link_pagamento: {
      type: DataTypes.TEXT,
      allowNull: true,
      comment: 'Link para pagamento (boleto, PIX, etc)'
    },
    // ✅ CAMPOS ADICIONAIS
    link_curto: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Link encurtado para compartilhamento'
    },
    codigo_barras: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Código de barras do boleto'
    },
    linha_digitavel: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Linha digitável do boleto'
    },
    qr_code: {
      type: DataTypes.TEXT,
      allowNull: true,
      comment: 'QR Code para PIX'
    },
    qr_code_base64: {
      type: DataTypes.TEXT,
      allowNull: true,
      comment: 'QR Code em base64'
    },
    observacoes: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    dados_mp: {
      type: DataTypes.JSONB,
      allowNull: true,
      comment: 'Dados completos retornados pelo Mercado Pago'
    },
    // ✅ NOVO CAMPO PARA URL DO COMPROVANTE
    comprovante_url: {
      type: DataTypes.TEXT,
      allowNull: true,
      comment: 'URL do comprovante oficial do Mercado Pago'
    },
    // ✅ CAMPOS DE JUROS (se ainda não existirem)
    valor_original: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Valor original sem juros (formatado)'
    },
    valor_original_numerico: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true,
      comment: 'Valor original sem juros (numérico)'
    },
    juros_total: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Total de juros (formatado)'
    },
    juros_total_numerico: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true,
      comment: 'Total de juros (numérico)'
    },
    taxa_juros: {
      type: DataTypes.DECIMAL(5, 2),
      allowNull: true,
      comment: 'Taxa de juros em percentual'
    },
    calculo_mp: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      comment: 'Se foi calculado pela API do Mercado Pago'
    },
    // ✅ NOVOS CAMPOS PARA SISTEMA DE PARCELAS
    parcela_atual: {
      type: DataTypes.INTEGER,
      defaultValue: 1,
      comment: 'Número da parcela atual (1, 2, 3...)'
    },
    pagamento_pai_id: {
      type: DataTypes.INTEGER,
      allowNull: true,
      comment: 'ID do pagamento principal (se for uma parcela)'
    },
    is_parcelado: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      comment: 'Se é um pagamento parcelado'
    },
    data_envio_proxima_parcela: {
      type: DataTypes.DATE,
      allowNull: true,
      comment: 'Data para envio da próxima parcela'
    },
    juros_mp: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true,
      comment: 'Juros calculados pelo Mercado Pago'
    },
    valor_com_juros: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true,
      comment: 'Valor total com juros do MP'
    },
    link_unico: {
      type: DataTypes.STRING,
      allowNull: true,
      unique: true,
      comment: 'Link único para esta parcela específica'
    },
    tenant_id: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: { model: 'tenants', key: 'id' }
    }
  }, {
    sequelize,
    modelName: 'Pagamento',
    tableName: 'pagamentos',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  });

  return Pagamento;
};

File: backend-go/internal/models/pagamento.go
package models

import (
	"time"

	"gorm.io/datatypes"
)

// Pagamento espelha a tabela `pagamentos` — cobranças avulsas de clientes.
// Migração de Mercado Pago → Asaas (ver 03-spec §"De-Para" e §"Campos a renomear").
// Campos MP (mp_preference_id, mp_payment_id, dados_mp, calculo_mp, juros_mp)
// foram substituídos pelos equivalentes Asaas/gateway-agnósticos.
// Tabela é escopada por tenant (tenant_id) — filtrada automaticamente pelos
// callbacks GORM via internal/tenant quando a query usa db.WithContext(ctx).
type Pagamento struct {
	ID        uint  `gorm:"primaryKey" json:"id"`
	ClienteID uint  `gorm:"column:cliente_id;index" json:"cliente_id"`
	CreatedBy *uint `gorm:"column:created_by;index" json:"created_by,omitempty"`
	TenantID  *uint `gorm:"column:tenant_id;index" json:"tenant_id,omitempty"`

	// Identificadores Asaas (substituem mp_preference_id/mp_payment_id).
	AsaasCustomerID *string `gorm:"column:asaas_customer_id" json:"asaas_customer_id,omitempty"`
	AsaasPaymentID  *string `gorm:"column:asaas_payment_id;index" json:"asaas_payment_id,omitempty"`

	Tipo   string `gorm:"column:tipo;default:universal" json:"tipo"` // boleto|pix|cartao|universal
	Status string `gorm:"column:status;default:pendente" json:"status"`
	// pendente|aprovado|rejeitado|cancelado|expirado|aguardando
	// (canônico recomendado: adotar vocabulário Asaas internamente e traduzir na saída)

	Titulo    string `gorm:"column:titulo" json:"titulo"`
	Descricao string `gorm:"column:descricao" json:"descricao"`

	// `valor` é STRING formatada (ex: "2.000,00") — legado do Node. Usar sempre
	// ValorNumerico (DECIMAL) para cálculos (gotcha 03-spec §2).
	Valor         string  `gorm:"column:valor" json:"valor"`
	ValorNumerico float64 `gorm:"column:valor_numerico" json:"valor_numerico"`

	Parcelas               int     `gorm:"column:parcelas;default:1" json:"parcelas"`
	ValorParcela            string  `gorm:"column:valor_parcela" json:"valor_parcela"`
	ValorParcelaNumerico    float64 `gorm:"column:valor_parcela_numerico" json:"valor_parcela_numerico"`

	WhatsappEnviado    bool       `gorm:"column:whatsapp_enviado;default:false" json:"whatsapp_enviado"`
	EmailEnviado       bool       `gorm:"column:email_enviado;default:false" json:"email_enviado"`
	DataEnvioWhatsapp  *time.Time `gorm:"column:data_envio_whatsapp" json:"data_envio_whatsapp,omitempty"`
	DataEnvioEmail     *time.Time `gorm:"column:data_envio_email" json:"data_envio_email,omitempty"`

	DataVencimento *time.Time `gorm:"column:data_vencimento" json:"data_vencimento,omitempty"`
	DataPagamento  *time.Time `gorm:"column:data_pagamento" json:"data_pagamento,omitempty"`

	// InvoiceURL substitui `link_pagamento`/`init_point` (MP) — Asaas `invoiceUrl`.
	InvoiceURL     string `gorm:"column:invoice_url" json:"invoice_url"`
	LinkCurto      string `gorm:"column:link_curto" json:"link_curto"`
	CodigoBarras   string `gorm:"column:codigo_barras" json:"codigo_barras"`
	LinhaDigitavel string `gorm:"column:linha_digitavel" json:"linha_digitavel"`
	QRCode         string `gorm:"column:qr_code" json:"qr_code"`
	QRCodeBase64   string `gorm:"column:qr_code_base64" json:"qr_code_base64"`

	Observacoes string `gorm:"column:observacoes" json:"observacoes"`

	// DadosGateway substitui `dados_mp` (JSONB) — payload bruto do gateway (Asaas).
	DadosGateway datatypes.JSON `gorm:"column:dados_gateway" json:"dados_gateway,omitempty"`

	// InvoiceURL cobre comprovante_url/receipt_url; TransactionReceiptURL é o
	// comprovante de transação quando disponível (Asaas `transactionReceiptUrl`).
	TransactionReceiptURL string `gorm:"column:transaction_receipt_url" json:"transaction_receipt_url,omitempty"`

	ValorOriginal         string   `gorm:"column:valor_original" json:"valor_original,omitempty"`
	ValorOriginalNumerico *float64 `gorm:"column:valor_original_numerico" json:"valor_original_numerico,omitempty"`
	JurosTotal            string   `gorm:"column:juros_total" json:"juros_total,omitempty"`
	JurosTotalNumerico    *float64 `gorm:"column:juros_total_numerico" json:"juros_total_numerico,omitempty"`
	TaxaJuros             *float64 `gorm:"column:taxa_juros" json:"taxa_juros,omitempty"`

	ParcelaAtual              *int       `gorm:"column:parcela_atual" json:"parcela_atual,omitempty"`
	PagamentoPaiID            *uint      `gorm:"column:pagamento_pai_id" json:"pagamento_pai_id,omitempty"`
	IsParcelado               bool       `gorm:"column:is_parcelado;default:false" json:"is_parcelado"`
	DataEnvioProximaParcela   *time.Time `gorm:"column:data_envio_proxima_parcela" json:"data_envio_proxima_parcela,omitempty"`

	ValorComJuros *float64 `gorm:"column:valor_com_juros" json:"valor_com_juros,omitempty"`

	LinkUnico string `gorm:"column:link_unico;uniqueIndex" json:"link_unico"`

	CreatedAt time.Time `gorm:"column:created_at" json:"created_at"`
	UpdatedAt time.Time `gorm:"column:updated_at" json:"updated_at"`
}

func (Pagamento) TableName() string { return "pagamentos" }

// Status canônicos (vocabulário local, traduzido de/para eventos Asaas na borda).
const (
	PagamentoStatusPendente   = "pendente"
	PagamentoStatusAprovado   = "aprovado"
	PagamentoStatusRejeitado  = "rejeitado"
	PagamentoStatusCancelado  = "cancelado"
	PagamentoStatusExpirado   = "expirado"
	PagamentoStatusAguardando = "aguardando"
)

// Tipos de cobrança suportados.
const (
	PagamentoTipoBoleto    = "boleto"
	PagamentoTipoPix       = "pix"
	PagamentoTipoCartao    = "cartao"
	PagamentoTipoUniversal = "universal"
)
